File: src/main/java/com/baidu/algorithm/bfs/_207_Course_Schedule.py
from collections import deque

class Solution(object):
    # 拓扑排序
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        if numCourses < 2:
            return True

        in_degree = [0] * numCourses
        adj = [[] for _ in xrange(numCourses)]

        for x, y in prerequisites:
            in_degree[x] += 1
            adj[y].append(x)

        dq = deque()
        for x in xrange(numCourses):
            if not in_degree[x]:
                dq.append(x)

        cnt = 0
        while dq:
            x = dq.popleft()
            cnt += 1
            for i in adj[x]:
                in_degree[i] -= 1
                if in_degree[i] == 0:
                    dq.append(i)

        return cnt == numCourses

File: src/main/java/com/baidu/algorithm/bfs/_207_Course_Schedule.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.bfs;

import java.util.*;

/**
 * _207_Course_Schedule
 *
 * @author xuhaoran01
 */
public class _207_Course_Schedule {

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if (numCourses < 2) {
            return true;
        }

        int[] inDegree = new int[numCourses];
        Map<Integer, List<Integer>> outs = new HashMap<>();
        for (int[] pr : prerequisites) {
            inDegree[pr[0]]++;
            if (outs.containsKey(pr[1])) {
                outs.get(pr[1]).add(pr[0]);
            } else {
                outs.put(pr[1], new ArrayList<>(Arrays.asList(pr[0])));
            }
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) {
                queue.add(i);
            }
        }

        int outCnt = 0;
        while (!queue.isEmpty()) {
            int t = queue.poll();
            outCnt++;

            List<Integer> out = outs.getOrDefault(t, new ArrayList<>());
            for (int i = 0; i < out.size(); i++) {
                inDegree[out.get(i)]--;
                if (inDegree[out.get(i)] == 0) {
                    queue.add(out.get(i));
                }
            }
        }

        return outCnt == numCourses;
    }

    public static void main(String[] args) {
        new _207_Course_Schedule().canFinish(3, new int[][]{{1, 0}, {2, 1}});
    }
}
